lib/ansible/cli/doc.py

```diff
@@ -172,22 +371,93 @@ def display_plugin_list(self, results):
         # display results
         DocCLI.pager("\n".join(text))
 
+    def _display_available_roles(self, list_json):
+        """Display all roles we can find with a valid argument specification.
+
+        Output is: fqcn role name, entry point, short description
+        """
+        roles = list(list_json.keys())
+        entry_point_names = set()
+        for role in roles:
+            for entry_point in list_json[role]['entry_points'].keys():
+                entry_point_names.add(entry_point)
+
+        max_role_len = 0
+        max_ep_len = 0
+
+        if roles:
+            max_role_len = max(len(x) for x in roles)
+        if entry_point_names:
+            max_ep_len = max(len(x) for x in entry_point_names)
+
+        linelimit = display.columns - max_role_len - max_ep_len - 5
+        text = []
+
+        for role in sorted(roles):
+            for entry_point, desc in iteritems(list_json[role]['entry_points']):
+                if len(desc) > linelimit:
+                    desc = desc[:linelimit] + '...'
+                text.append("%-*s %-*s %s" % (max_role_len, role,
+                                              max_ep_len, entry_point,
+                                              desc))
+
+        # display results
+        DocCLI.pager("\n".join(text))
+
+    def _display_role_doc(self, role_json):
+        roles = list(role_json.keys())
+        text = []
+        for role in roles:
+            text += self.get_role_man_text(role, role_json[role])
+
+        # display results
+        DocCLI.pager("\n".join(text))
+
     def run(self):
 
         super(DocCLI, self).run()
 
+        basedir = context.CLIARGS['basedir']
         plugin_type = context.CLIARGS['type']
         do_json = context.CLIARGS['json_format']
+        roles_path = context.CLIARGS['roles_path']
+
+        if basedir:
+            AnsibleCollectionConfig.playbook_paths = basedir
+
+            # Add any 'roles' subdir in playbook dir to the roles search path
+            subdir = os.path.join(basedir, "roles")
+            if os.path.isdir(subdir):
+                roles_path = (subdir,) + roles_path
```

Comment:
There are some quirks with the roles search path. If not found in the `roles` subdir, the basedir is also searched. In a collection if you have a `.` in the role name it's another directory layer - `geerlingguy.docker` in a collection should refer to `geerlingguy/docker/` not `geerlingguy.docker/`. `ansible-doc` doesn't need all of it, but it might be worth refactoring [_load_role_path](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/definition.py#L137) to move the search path resolution (and maybe the role path loading?) to common code so everything stays in sync.

Reply:
Hrm, I'm probably being dense here, so you may have to help me out. Not sure I follow the statement...

  `If not found in the roles subdir, the basedir is also searched.`

Is my code doing that, or code elsewhere? I don't see where that is done, and doesn't seem like correct behavior for finding roles.

The logic I've added here searches for roles in:

- DEFAULT_ROLES_PATH
- the `roles` subdir of whatever value is passed to `--playbook-dir` (which is `basedir` in this portion of the code)
- the value of `--roles-path`
- the `roles` subdir of any collections that are found.

Failing to see where the collection bit you discussed fits in, too.   :(

Reply:
FYI, the collection bit fits in here https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/definition.py#L159

Reply:
I've added searching the playbook basedir as a last-resort in the latest commits. Thanks for seeing that undocumented behavior! I totally missed it.

After looking through the code in _load_role_path() and related bits, there is a lot of commonality, but there is also a lot of stuff happening there that ansible-doc doesn't necessarily need to do. We definitely should work to consolidate this code, but I think attempting to do so here would make this already large change much more complex. I think that should be a separate project (to be done soon, hopefully).

Reply:
Agreed- trying to refactor/share the common bits is probably out of scope for this PR. If we end up doing the `importlib.resources` stuff though, that's probably the place to step back and take a look at where we can share a bunch of this stuff between the runtime and docs generation, and try to come up with a more coherent generalized API.